src/main/java/marquez/core/services/JobService.java

```diff
@@ -0,0 +1,224 @@
+          jobVersionDAO.insert(JobService.createJobVersion(existingJobWithNewUri));
+          return existingJobWithNewUri;
+        } else {
+          return existingJob;
+        }
+      }
+    } catch (UnableToExecuteStatementException e) {
+      String err = "failed to create new job";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public List<Job> getAllJobsInNamespace(String namespace) throws UnexpectedException {
+    try {
+      return this.jobDAO.findAllInNamespace(namespace);
+    } catch (UnableToExecuteStatementException e) {
+      log.error("caught exception while fetching jobs in namespace ", e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public List<JobVersion> getAllVersionsOfJob(String namespace, String jobName)
+      throws UnexpectedException {
+    try {
+      return this.jobVersionDAO.find(namespace, jobName);
+    } catch (UnableToExecuteStatementException e) {
+      log.error("caught exception while fetching versions of job", e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public Optional<JobVersion> getLatestVersionOfJob(String namespace, String jobName)
+      throws UnexpectedException {
+    try {
+      return Optional.ofNullable(jobVersionDAO.findLatest(namespace, jobName));
+    } catch (UnableToExecuteStatementException e) {
+      String err = "error fetching latest version of job";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public JobRun updateJobRunState(UUID jobRunID, JobRunState.State state)
+      throws UnexpectedException {
+    try {
+      this.jobRunDAO.updateState(jobRunID, JobRunState.State.toInt(state));
+      return this.jobRunDAO.findJobRunById(jobRunID);
+    } catch (UnableToExecuteStatementException e) {
+      String err = "error updating job run state";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public Optional<JobRun> getJobRun(UUID jobRunID) throws UnexpectedException {
+    try {
+      return Optional.ofNullable(this.jobRunDAO.findJobRunById(jobRunID));
+    } catch (UnableToExecuteStatementException e) {
+      String err = "error fetching job run";
+      log.error(err, e);
+      throw new UnexpectedException();
+    }
+  }
+
+  public JobRun createJobRun(
+      String namespaceName,
+      String jobName,
+      String runArgsJson,
+      Timestamp nominalStartTime,
+      Timestamp nominalEndTime)
+      throws UnexpectedException {
+    try {
+      String runArgsDigest = null;
+      RunArgs runArgs = null;
+      if (null == jobDAO.findByName(namespaceName, jobName)) {
+        String err =
+            String.format(
+                "unable to find job <ns='%s', job name='%s'> to create job run",
+                namespaceName, jobName);
+        log.error(err);
+        throw new UnexpectedException();
+      }
+      Optional<JobVersion> latestJobVersion = getLatestVersionOfJob(namespaceName, jobName);
```

Comment:
look into `orThrow` to just raise exception automatically

Reply:
Looked into `orElseThrow`, but there's an additional heap allocation of an `Optional<Job>` just to get the syntactic sugar. It does look nicer (`Optional.ofNullable(jobDAO.findByName(namespaceName, jobName)).orElseThrow(...)`), but I'm not sure if it's worthwhile unless the DAOs start returning `Optional`s. Maybe we can look into that for the future:

https://stackoverflow.com/questions/31395271/register-optional-for-jdbi

Reply:
#185